### PR TITLE
[usability][rfc] ray status show demand summary by default

### LINKED ISSUE #15867: [core/autoscaler] Don't show placement-group specific resource demands in status
### What is the problem?

Can we avoid printing the placement group specific info when printing resources (and toggle it or something)?

For example:

```
(raybuild) ➜  tune git:(master) ✗ ray status
======== Cluster status: 2021-05-13 00:19:58.414667 ========
Node status
------------------------------------------------------------
 1 node(s) with resources: {'object_store_memory': 15357854515.0, 'CPU': 16.0, 'node:192.168.1.115': 1.0, 'memory': 30715709031.0}

Resources
------------------------------------------------------------
Usage:
 0.0/16.0 CPU
 0.00/28.606 GiB memory
 0.00/14.303 GiB object_store_memory

Demands:
 {'CPU_group_4a82a217aadd8326a3a49f02700ac5c2': 0.001, 'CPU_group_0_4a82a217aadd8326a3a49f02700ac5c2': 0.001}: 1+ pending tasks/actors
 {'CPU_group_0_811ff50281b27f77fd4c0836c5420248': 0.001, 'CPU_group_811ff50281b27f77fd4c0836c5420248': 0.001}: 1+ pending tasks/actors
 {'CPU_group_90d343fba962c84bd9cdafa8ab9dcef7': 0.001, 'CPU_group_0_90d343fba962c84bd9cdafa8ab9dcef7': 0.001}: 1+ pending tasks/actors
 {'CPU_group_0_fd4f17e294e92436ecd1b9a647f71411': 0.001, 'CPU_group_fd4f17e294e92436ecd1b9a647f71411': 0.001}: 1+ pending tasks/actors
```

Also occurs when "resources not available" message is printed.

Instead, we should aggregate these resources (e.g., ("CPUs reserved for placement groups")).

### PR DESCRIPTION
**Pull request by scv119**
## Why are these changes needed?

currently `ray summary` shows placement group related demands information which is a bit hard to understand:
```
ray status
======== Cluster status: 2021-08-17 01:15:02.033448 ========
Node status
------------------------------------------------------------
 1 node(s) with resources: {'node:192.168.86.195': 1.0, 'CPU_group_625ace126f848864c46f50dced5e0ef7': 16.0, 'CPU_group_0_625ace126f848864c46f50dced5e0ef7': 16.0, 'CPU': 16.0, 'object_store_memory': 10202942668.0, 'memory': 20405885339.0}

Resources
------------------------------------------------------------
Usage:
 16.0/16.0 CPU
 0.0/16.0 CPU_group_0_625ace126f848864c46f50dced5e0ef7
 10.0/16.0 CPU_group_625ace126f848864c46f50dced5e0ef7
 0.00/19.004 GiB memory
 0.00/9.502 GiB object_store_memory

Demands:
 {'CPU_group_625ace126f848864c46f50dced5e0ef7': 10.0}: 1+ pending tasks/actors
 {'CPU': 100.0}: 1+ pending tasks/actors
 {'memory': 2040588533900.0} * 1 (STRICT_PACK): 2+ pending placement groups
 {'GPU': 2.0} * 1, {'extra_resource': 2.0} * 1 (STRICT_PACK): 1+ pending placement groups
 {'CPU': 16.0} * 1, {'memory': 2040588533900.0} * 1 (STRICT_PACK): 1+ pending placement groups
 {'CPU': 1}: 3+ from request_resources()
 {'GPU': 1, 'CPU': 4}: 1+ from request_resources()
```

in this PR, we create a summary of resource demands and group by tasks/actors, placement groups and resource requirements, as following:

```
ray status
======== Cluster status: 2021-08-17 01:15:12.045287 ========
Node status
------------------------------------------------------------
 1 node(s) with resources: {'node:192.168.86.195': 1.0, 'object_store_memory': 10202942668.0, 'CPU_group_0_625ace126f848864c46f50dced5e0ef7': 16.0, 'CPU': 16.0, 'CPU_group_625ace126f848864c46f50dced5e0ef7': 16.0, 'memory': 20405885339.0}

Resources
------------------------------------------------------------
Usage:
 16.0/16.0 CPU
 0.00/19.004 GiB memory
 0.00/9.502 GiB object_store_memory

Demands:
 2 pending tasks/actors:
  110.0 CPU
 4 pending placement groups:
  16.0 CPU
  2.0 GPU
  2.0 extra_resource
  5701.339 GiB memory
 4 pending resource requirements:
  7.0 CPU
  1.0 GPU
```

## Related issue number

#15867

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by wuisawesome** [comment]
This looks great to me. @richardliaw do you have any ux comments?

**Comment by rkooo567** [comment]
I think we should also format the usage?

**Comment by richardliaw** [comment]
Usage formatting +1

**Comment by scv119** [comment]
actually it's a bit confusing to show per replacement group resource usage; mostly because the capacity adds up exceeds the physical capacity.
so instead we just not showing per replacement group resource unless `ray status --verbose` is specified.

**Comment by richardliaw** [comment]
That is really nice!

**Review by ericl — CHANGES_REQUESTED** [review]
This is actually losing a lot of the important information right? Can we scope it to just to the placement groups or otherwise avoid losing info via aggregation?

**Comment by scv119** [comment]
@ericl  so with this PR if we run `ray status --verbose` it works as the same as the previous version. does that work or we should limit the change to placement group only?

**Comment by richardliaw** [comment]
my 2c is that I like the direction of hiding those placement group names - it's quite hard to parse otherwise.

To avoid loss of info, we should at least have a summary indicator about placement groups usage (how many placements are created), and a hint for the user as to what they need to do to see the placement group usage.

**Comment by ericl** [comment]
Hmm I was thinking the main fix would be to "handle placement groups" properly in the original display:

For example:
```
CPU: 16.0/16.0 (8.0 reserved in placement groups)
```

```
{"CPU": 1}: 10+ pending tasks/actors (3+ using placement groups)
```

**Comment by scv119** [comment]
just so that we are agree on the same design, is this the new output we are looking for? (in form of diff):

```
 Resources
 ------------------------------------------------------------
 Usage:
- 16.0/16.0 CPU
- 0.0/16.0 CPU_group_0_625ace126f848864c46f50dced5e0ef7
- 10.0/16.0 CPU_group_625ace126f848864c46f50dced5e0ef7
+ 16.0/16.0 CPU (10.0 reserved in placement groups)
  0.00/19.004 GiB memory
  0.00/9.502 GiB object_store_memory

 Demands:
- {'CPU_group_625ace126f848864c46f50dced5e0ef7': 10.0}: 1+ pending tasks/actors
- {'CPU': 100.0}: 1+ pending tasks/actors
+ {'CPU': 100.0}: 2+ pending tasks/actors (1+ using placement groups)
  {'memory': 2040588533900.0} * 1 (STRICT_PACK): 2+ pending placement groups
  {'GPU': 2.0} * 1, {'extra_resource': 2.0} * 1 (STRICT_PACK): 1+ pending placement groups
  {'CPU': 16.0} * 1, {'memory': 2040588533900.0} * 1 (STRICT_PACK): 1+ pending placement groups
```

**Comment by ericl** [comment]
Diff LGTM

**Comment by scv119** [comment]
actually the tasks/actors are not grouped so here is the diff after update:
```
 Resources
 ------------------------------------------------------------
 Usage:
- 16.0/16.0 CPU
- 0.0/16.0 CPU_group_0_625ace126f848864c46f50dced5e0ef7
- 10.0/16.0 CPU_group_625ace126f848864c46f50dced5e0ef7
+ 16.0/16.0 CPU (10.0 reserved in placement groups)
  0.00/19.004 GiB memory
  0.00/9.502 GiB object_store_memory

 Demands:
- {'memory_group_625ace126f848864c46f50dced5e0ef7': 10.0, 'CPU_group_625ace126f848864c46f50dced5e0ef7': 10000.0}: 1+ pending tasks/actors
- {'GPU_group_4a133bab2f93fe4a104533be2d51f85a': 10.0}: 5+ pending tasks/actors
+ {'memory': 10.0, 'CPU': 10000.0}: 1+ pending tasks/actors (using placement group 625ace126f848864c46f50dced5e0ef7)
+ {'GPU': 10.0}: 5+ pending tasks/actors (using placement group 4a133bab2f93fe4a104533be2d51f85a)
  {'CPU': 100.0}: 1+ pending tasks/actors
  {'CPU': 30.0, 'GPU': 2.0}: 3+ pending tasks/actors
  {'memory': 2040588533900.0} * 1 (STRICT_PACK): 2+ pending placement groups
  {'GPU': 2.0} * 1, {'extra_resource': 2.0} * 1 (STRICT_PACK): 1+ pending placement groups
  {'CPU': 16.0} * 1, {'memory': 2040588533900.0} * 1 (STRICT_PACK): 1+ pending placement groups
```

**Comment by ericl** [comment]
> + {'memory': 10.0, 'CPU': 10000.0}: 1+ pending tasks/actors (using placement group 625ace126f848864c46f50dced5e0ef7)

Question for this, if there are many placement groups would the line get too long? Might be better to just have a count.

**Comment by scv119** [comment]
hmm currently if we have many placement groups they will be showed in different lines like following:
```
 {'memory': 10.0, 'CPU': 10.0}: 1+ pending tasks/actors (using placement group 625ace126f848864c46f50dced5e0ef7)
 {'memory': 12.0, 'CPU': 8.0}: 1+ pending tasks/actors (using placement group 4a133bab2f93fe4a104533be2d51f85a)
```
are you suggesting we showing this?
```
 {'memory': 10.0, 'CPU': 10.0}: 1+ pending tasks/actors (1+ using placement group)
 {'memory': 12.0, 'CPU': 8.0}: 1+ pending tasks/actors (1+ using placement group)
```

**Comment by ericl** [comment]
If the two demand bundles are the same size (say 1 CPU each), which is the common case, I do expect them to be aggregated, for example:

bad:
```
 {'CPU': 1.0}: 7+ pending tasks
 {'CPU': 1.0}: 1+ pending tasks/actors (using placement group 625ace126f848864c46f50dced5e0ef7)
 {'CPU': 1.0}: 1+ pending tasks/actors (using placement group 4a133bab2f93fe4a104533be2d51f85a)
 {'CPU': 1.0}: 1+ pending tasks/actors (using placement group 123451234512342893472937498324)
```

good:
```
 {CPU': 1.0}: 10+ pending tasks/actors (3+ using placement groups)
```

**Comment by wuisawesome** [comment]
Actually a higher level question about that: shouldn't we not show those demands at all? they're an implementation detail of placement groups (i.e. the autoscaler will never be able to fulfill those demands, they should only show up as demands if the placement group itself can't be placed).

**Comment by ericl** [comment]
We are only showing demands if the placement group / task / actor isn't
placed (pending); if everything's placed the list should already be empty.

Eric

File: python/ray/autoscaler/_private/util.py [hunk 1]
```diff
@@ -447,27 +451,112 @@ def format_pg(pg):
...
-        line = f" {bundle}: {count}+ pending tasks/actors"
+    for bundle, count in bundle_demand.items():
+        line = f" {dict(bundle)}: {count}+ pending tasks/actors"
+        if bundle in pg_bundle_demand:
+            line += f" ({pg_bundle_demand[bundle]}+ using placement groups)"
         demand_lines.append(line)
+    return demand_lines
+
+
+def get_demand_report(lm_summary):
```
Comment (ericl) [inline python/ray/autoscaler/_private/util.py:555]: Add types?

File: python/ray/autoscaler/_private/util.py [hunk 2]
```diff
@@ -447,27 +451,112 @@ def format_pg(pg):
...
+                line = line + f" ({used_in_pg:.2f} GiB reserved" \
+                    + " in placement groups)"
         usage_lines.append(line)
     usage_report = "\n".join(usage_lines)
     return usage_report
 
 
-def get_demand_report(lm_summary):
+def format_resource_demand_summary(resource_demand: List[Any]) -> List[str]:
+    def filter_placement_group_from_bundle(bundle):
```
Comment (ericl) [inline python/ray/autoscaler/_private/util.py:516]: types?

**Review by ericl — APPROVED** [review]
The output LGTM

File: python/ray/autoscaler/_private/load_metrics.py [hunk 3]
```diff
@@ -23,6 +22,47 @@
 ])
 
 
+def add_resources(dict1: Dict[str, float],
```
Comment (scv119) [inline python/ray/autoscaler/_private/load_metrics.py:25]: move from util to here to solve circular dependency when using the LoadMetricsSummary type
